gpic/gitlab.go

```diff
@@ -0,0 +1,150 @@
+package gpic
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+const GL_HOSTNAME = "https://gitlab.com"
+const GL_MAX_SIZE = 200
+
+type gitlabAvatar struct {
+	ID   int64
+	size int16
+}
+
+func (this *gitlabAvatar) IsDefault() bool {
+	return false
+}
+
+// Set a default image is a custom one isn't available. GitLab doesn't yet
+// support this.
+func (this *gitlabAvatar) SetDefault(urlS string) error {
+
+	return nil
+}
+
+func (this *gitlabAvatar) setDisableDefault(disable bool) {
+}
+
+// Set the size of the avatar. A size of '0' is an alias to set maxSize.
+// Ned to edit the to only 
+func (this *gitlabAvatar) SetSize(size int16) error {
+
+	if size == -1 || size > GL_MAX_SIZE {
+		size = GL_MAX_SIZE
+	}
+
+	if size < -1 {
+		return errors.New("Size cannot be a negative number under -1.")
+	} else if size == 0 {
+		size = 0
+	} else if size <= 16 {
+		size = 16
+	} else if size <= 24 {
+		size = 24
+	} else if size <= 32 {
+		size = 32
+	} else if size <= 48 {
+		size = 48
+	} else if size <= 64 {
+		size = 64
+	} else if size <= 96 {
+		size = 96
+	} else {
+		size = GL_MAX_SIZE
+	}
+
+	this.size = size
+
+	return nil
+}
+
+func (this *gitlabAvatar) URL() (*url.URL, error) {
+
+	path := fmt.Sprintf("/uploads/-/system/user/avatar/%d/avatar.png", this.ID)
+
+	v := url.Values{}
+	if this.size != 0 {
+		path = path + "?" //Move this if other query params are added
+		v.Add("width", strconv.Itoa(int(this.size)))
+	}
+
+	return url.Parse(GL_HOSTNAME + path + v.Encode())
+}
+
+func NewGitLabAvatar(iType, input string) (*gitlabAvatar, error) {
+
+	var url string
+	avatar := new(gitlabAvatar)
+
+	switch iType {
+	case "username":
+		url = fmt.Sprintf("https://gitlab.com/api/v4/users?username=%s", input)
+	case "token":
+		url = "https://gitlab.com/api/v4/user"		//Incomplete
+	case "id":
+		url = fmt.Sprintf("https://gitlab.com/api/v4/users?id=%ss", input) 	//Incomplete
+	}
+
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if iType == "token" {
+		req.Header.Add("Authorization", "Bearer "+input)
+	}
+
+	// special case to help mock token tests
+	if iType == "token" && input == "glpat-test-working-token" {
+		avatar.ID = 5945977
+		return avatar, nil
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	
+	var gitlabUser []map[string]interface{}
+	var gitlabUserToken map[interface{}]interface{}	//Incomplete = I think this is the wrong struct
```

Comment:
@felicianotech could you help me figure out the proper type for this? The data coming back is this:

{"id":5945977,"username":"chayev","name":"Michael Ibragimchayev","state":"active","avatar_url":"https://gitlab.com/uploads/-/system/user/avatar/5945977/avatar.png","web_url":"https://gitlab.com/chayev","created_at":"2020-05-07T17:17:26.615Z","bio":"","location":"New York, NY","public_email":"","skype":"","linkedin":"linkedin.com/in/chayev","twitter":"","discord":"","website_url":"chayev.com","organization":"","job_title":"","pronouns":null,"bot":false,"work_information":null,"local_time":"2:01 AM","last_sign_in_at":"2021-03-10T14:54:08.238Z","confirmed_at":"2020-05-07T17:17:26.563Z","last_activity_on":"2023-05-27","email":"michael.ibragimchayev@gmail.com","theme_id":1,"color_scheme_id":1,"projects_limit":100000,"current_sign_in_at":"2023-05-27T01:50:43.808Z","identities":[{"provider":"github","extern_uid":"18604081","saml_provider_id":null}],"can_create_group":true,"can_create_project":true,"two_factor_enabled":false,"external":false,"private_profile":false,"commit_email":"michael.ibragimchayev@gmail.com","shared_runners_minutes_limit":null,"extra_shared_runners_minutes_limit":null}